KFSPTS-8653 handle accounting info validation for personal corporate card charges

db483 opened the pull request.

cah292 commented on src/main/java/edu/cornell/kfs/concur/service/ConcurAccountValidationService.java:
```diff
@@ -7,4 +7,5 @@
     
     ValidationResult validateConcurAccountInfo(ConcurAccountInfo concurAccountInfo);
 
+    ValidationResult validateConcurAccountInfoObjectCodeNotRequired(ConcurAccountInfo concurAccountInfo);
```
Can you please clarify the intended object code validation? In this code (and as per your recent user story comment), you're bypassing the object code validation altogether for the personal corp card lines. However, as per the requirements on the user story, it sounds as though they should be validated using a parameter-based object code and an empty sub-object code instead. If the story's acceptance criteria are out-of-date, then can you please update them?

db483 replied:
The acceptance criteria is out of date. I will update it.

For personal corp card expenses there is no object code value in the report details so only the rest of the accounting string will be validated.

cah292 commented on src/main/java/edu/cornell/kfs/concur/service/impl/ConcurAccountValidationServiceImpl.java:
```diff
@@ -40,28 +39,46 @@ public ValidationResult validateConcurAccountInfo(ConcurAccountInfo concurAccoun
 
-    public ValidationResult checkAccountingString(String chartOfAccountsCode, String accountNumber, String subAccountNumber, String objectCode, String subObjectCode, String projectCode) {
-        ValidationResult requiredFieldsValidationResult = checkRequiredAccountInfo(chartOfAccountsCode, accountNumber, objectCode);
+    public ValidationResult checkAccountingString(String chartOfAccountsCode, String accountNumber, String subAccountNumber, String objectCode, String subObjectCode, String projectCode, boolean objectCodeRequired) {
```
If you added the ConcurBatchUtilityService method suggested from my earlier comment, you might be able to refactor this class to avoid these boolean method arguments.

db483 replied:
That would be nice indeed, but it cannot determine if it is personal corp card expense based only on the accounting string information.

cah292 replied:
Ok, that makes sense.

db483 commented on src/main/java/edu/cornell/kfs/concur/businessobjects/ConcurAccountInfo.java:
```diff
@@ -10,9 +10,9 @@
     protected String subObjectCode;
     protected String projectCode;
     protected String orgRefId;
+    protected boolean forPersonalCorporateCardExpense;
```
There is no other information on the ConcurAccountInfo that can be used to determine if it is used for personal corp card. Adding a boolean value to determine that seemed like a reasonable approach.

cah292 replied:
Ok, thanks for clarifying.

db483 commented on src/main/java/edu/cornell/kfs/concur/ConcurConstants.java:
```diff
@@ -108,5 +108,8 @@
     }
+    
+    public static final String YES = "Y";
+    public static final String NO = "N";
```
will switch to using those.

db483 commented on src/main/java/edu/cornell/kfs/concur/rest/xmlObjects/ExpenseEntryDTO.java:
```diff
@@ -15,9 +15,12 @@
     private List<ItemizationEntryDTO> itemizationsList;
-
-    @XmlElement(name = "OrgUnit6")
-    private String orgUnit6;
```
This is not used in the code so I removed it.

db483 commented:
Hi Chad,

I have updated the code and also added some changes based on some feedback from Sherry.

I have also updated the acceptance criteria on the user story.

Please review again. Thank you.
